core/iwasm/common/wasm_shared_memory.c

```diff
@@ -87,6 +92,62 @@ search_module(WASMModuleCommon *module)
     return NULL;
 }
 
+static void
+wait_map_address_count_callback(void *key, void *value,
+                                void *p_total_elem_count)
+{
+    *(uint32 *)p_total_elem_count = *(uint32 *)p_total_elem_count + 1;
+}
+
+static void
+create_list_of_waiter_addresses(void *key, void *value, void *user_data)
+{
+    AtomicWaitAddressArgs *data = (AtomicWaitAddressArgs *)user_data;
+    data->addr[data->index++] = key;
+}
+
+void
+notify_stale_threads_on_exception(WASMModuleInstanceCommon *module_inst)
+{
+    AtomicWaitAddressArgs *args = NULL;
+    uint32 i = 0, total_elem_count = 0;
+    if (!(args = wasm_runtime_malloc(sizeof(AtomicWaitAddressArgs)))) {
+        LOG_ERROR("failed to create atomic wait address args node");
+        return;
+    }
+
+    os_mutex_lock(&shared_memory_list_lock);
+
+    /* count number of addresses in wait_map */
+    bh_hash_map_traverse(wait_map, wait_map_address_count_callback,
+                         (void *)&total_elem_count);
+
+    /* allocate memory */
+    uint64 total_size = offsetof(AtomicWaitAddressArgs, addr)
+                        + sizeof(void *) * total_elem_count;
+
+    if (!(args->addr = wasm_runtime_malloc((uint32)total_size))) {
+        LOG_ERROR("failed to allocate memory for list of atomic wait addresses "
```

Comment:
You don't free the memory allocated for `args` if we reach here (although I still advise you to allocate it on allocate it on the stack instead).